src/infrastructure/jira/jira-service.ts

```diff
@@ -132,6 +132,21 @@ class JiraService {
 		{ url, displayName }: AtlassianDesign,
 		connectInstallation: ConnectInstallation,
 	): Promise<void> => {
+		const setIssueProperty = async (value: string): Promise<void> => {
```

Comment:
**suggestion:** The logic of this property update is quite complicated, and the implementation naturally gets complicated as well. It seems we could use a few tricks to make it shorter, more linear and easier to understand:

- Combine paths when the property is missing and when it has an an unexpected value.
- Use an early return when a property already contains a target URL.

Here is how it could look like (needs to be tested :)):

```ts
	updateAttachedDesignUrlV2IssueProperty = async (
		issueIdOrKey: string,
		{ url, displayName }: AtlassianDesign,
		connectInstallation: ConnectInstallation,
	): Promise<void> => {
		let newValue: AttachedDesignUrlV2IssuePropertyValue[];

		try {
			const response = await jiraClient.getIssueProperty(
				issueIdOrKey,
				propertyKeys.ATTACHED_DESIGN_URL_V2,
				connectInstallation,
			);

			const currentValue = JSON.parse(ensureString(response.value)) as unknown;

			assertSchema<AttachedDesignUrlV2IssuePropertyValue[]>(
				currentValue,
				ATTACHED_DESIGN_URL_V2_VALUE_SCHEMA,
			);

			if (currentValue.some((value) => value.url === url)) return;

			newValue = [...currentValue, { url, name: displayName }];
		} catch (error: unknown) {
			if (
				error instanceof JiraClientNotFoundError ||
				error instanceof SchemaValidationError ||
				error instanceof TypeError ||
				error instanceof SyntaxError
			) {
				newValue = [];
			} else {
				throw error;
			}
		}

		await jiraClient.setIssueProperty(
			issueIdOrKey,
			propertyKeys.ATTACHED_DESIGN_URL_V2,
			this.superStringify(newValue),
			connectInstallation,
		);
	};
```

We could simplify this even further if we introduce an additional utility function for parsing JSON of the expected schema (e.g., `parseJsonOfSchema`):

```ts
	updateAttachedDesignUrlV2IssueProperty = async (
		issueIdOrKey: string,
		{ url, displayName }: AtlassianDesign,
		connectInstallation: ConnectInstallation,
	): Promise<void> => {
		let newValue: AttachedDesignUrlV2IssuePropertyValue[];

		try {
			const response = await jiraClient.getIssueProperty(
				issueIdOrKey,
				propertyKeys.ATTACHED_DESIGN_URL_V2,
				connectInstallation,
			);

			const currentValue = parseJsonOfSchema< // A new utility function that returns a parsed object or throw `SchemaValidationError`.
				AttachedDesignUrlV2IssuePropertyValue[]
			>(response.value, ATTACHED_DESIGN_URL_V2_VALUE_SCHEMA);

			if (currentValue.some((value) => value.url === url)) return;

			newValue = [...currentValue, { url, name: displayName }];
		} catch (error: unknown) {
			if (
				error instanceof JiraClientNotFoundError ||
				error instanceof SchemaValidationError
			) {
				newValue = [];
			} else {
				throw error;
			}
		}

		await jiraClient.setIssueProperty(
			issueIdOrKey,
			propertyKeys.ATTACHED_DESIGN_URL_V2,
			this.superStringify(newValue),
			connectInstallation,
		);
	};
```

Reply:
Great suggestion, looks much easier to follow 🙂. I'll add the parsing util and refactor.